osu.Game.Rulesets.Osu.Tests/LegacyMainCirclePieceTest.cs

```diff
@@ -0,0 +1,97 @@
+                // available textures
+                new[] { @"hitcircle", @"hitcircleoverlay" },
+                // priority lookup
+                null,
+                // expected circle and overlay
+                @"hitcircle", @"hitcircleoverlay",
+            },
+            // custom priority lookup
+            new object?[]
+            {
+                new[] { @"hitcircle", @"hitcircleoverlay", @"sliderstartcircle", @"sliderstartcircleoverlay" },
+                @"sliderstartcircle",
+                @"sliderstartcircle", @"sliderstartcircleoverlay",
+            },
+            // when no sprites are available for the specified prefix, fall back to "hitcircle"/"hitcircleoverlay".
+            new object?[]
+            {
+                new[] { @"hitcircle", @"hitcircleoverlay" },
+                @"sliderstartcircle",
+                @"hitcircle", @"hitcircleoverlay",
+            },
+            // when a circle is available for the specified prefix but no overlay exists, no overlay is displayed.
+            new object?[]
+            {
+                new[] { @"hitcircle", @"hitcircleoverlay", @"sliderstartcircle" },
+                @"sliderstartcircle",
+                @"sliderstartcircle", null
+            },
+            // when no circle is available for the specified prefix but an overlay exists, the overlay is ignored.
+            new object?[]
+            {
+                new[] { @"hitcircle", @"hitcircleoverlay", @"sliderstartcircleoverlay" },
+                @"sliderstartcircle",
+                @"hitcircle", @"hitcircleoverlay",
+            }
+        };
+
+        [TestCaseSource(nameof(texture_priority_cases))]
+        public void TestTexturePriorities(string[] textureFilenames, string priorityLookup, string? expectedCircle, string? expectedOverlay)
+        {
+            Sprite? circleSprite = null;
+            Sprite? overlaySprite = null;
+
+            AddStep("load circle piece", () =>
+            {
+                var skin = new Mock<ISkinSource>();
+
+                skin.Setup(s => s.GetTexture(It.IsAny<string>())).CallBase();
+                skin.Setup(s => s.GetTexture(It.IsIn(textureFilenames), It.IsAny<WrapMode>(), It.IsAny<WrapMode>()))
+                    .Returns((string componentName, WrapMode _, WrapMode __) => new Texture(1, 1) { AssetName = componentName });
+
+                Child = new DependencyProvidingContainer
+                {
+                    CachedDependencies = new (Type, object)[] { (typeof(ISkinSource), skin.Object) },
+                    Child = new LegacyMainCirclePiece(priorityLookup, false),
+                };
+
+                var sprites = this.ChildrenOfType<Sprite>().Where(s => s.Texture.AssetName != null).DistinctBy(s => s.Texture.AssetName).ToArray();
+                Debug.Assert(sprites.Length <= 2);
+
+                circleSprite = sprites.ElementAtOrDefault(0);
+                overlaySprite = sprites.ElementAtOrDefault(1);
```

Comment:
I dunno about doing this, it's pretty brittle. `OverlayLayer` is already exposed as protected, so exposing `hitCircleSprite` as protected too and using a test-specific subclass is potentially an alternative option here.